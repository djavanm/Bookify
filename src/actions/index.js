export const search = query => ({
  type: 'SEARCH_BOOKS',
  query
});
