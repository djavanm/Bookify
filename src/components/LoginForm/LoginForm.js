import React, { Component } from 'react';

class LoginForm extends Component {
  constructor() {
    super()
    this.state = {
      existingUser: true,
      name: '',
      email: '',
      password: ''
    }
  }
  render() {
    const { existingUser } = this.state;
    return (
      <form>hi</form>
    )
  }
}




export default LoginForm;
