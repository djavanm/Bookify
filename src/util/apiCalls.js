export const fetchOnLoad = () => {
  return fetch('https://itunes.apple.com/search?term=audiobooks&limit=10')
    .then(response => {
      if(!response.ok) {
        throw Error('ERROR')
      }
      return response.json()
    })
    .catch(error => Error(error.message))
}

export const getBooks = (query) => {
  return fetch(`https://itunes.apple.com/search?media=audiobook&term=${query}`)
    .then(response => {
      if(!response.ok) {
        throw Error('ERROR')
      }
      return response.json()
    })
    .catch(error => Error(error.message))
}
