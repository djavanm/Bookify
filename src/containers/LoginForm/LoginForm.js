import React, { Component } from 'react';
import { createUser } from '../../util/apiCalls';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { setUser, setFavorites, setGenres } from '../../actions';
import { getFavorites } from '../../util/apiCalls';
import PropTypes from 'prop-types';


export class LoginForm extends Component {
  constructor() {
    super()
    this.state = {
      existingUser: true,
      name: '',
      email: '',
      password: '',
      error: ''
    }
  };

  toggleExisting = () => {
    this.setState({existingUser: !this.state.existingUser})
  };

  handleChange = (e) => {
    this.setState({[e.target.name]: e.target.value})
  };

  submitUser = (e) => {
    e.preventDefault();
    const { setUser, setFavorites, setGenres } = this.props;
    const { name, email, password } = this.state;
    const newUser = {
      name,
      email,
      password
    };
    this.clearFields();
    createUser(newUser, 'users')
      .then(data => data.id ? setUser(data) : null)
      .then(data => data.foundUser.id ? getFavorites(data.foundUser.id) : null)
      .then(data => setFavorites(data.favorites))
      .then(data => setGenres(data.foundFavorites))
      .catch(error => this.setState({error: 'Email has already been used'}))
  };

  loginUser = (e) => {
    e.preventDefault();
    const { setUser, setFavorites, setGenres } = this.props;
    const { email, password } = this.state;
    const newUser = {
      email,
      password
    };
    this.clearFields();
    createUser(newUser, 'login')
      .then(data => data.id ? setUser(data) : null )
      .then(data => data.foundUser.id ? getFavorites(data.foundUser.id) : null)
      .then(data => setFavorites(data.favorites))
      .then(data => setGenres(data.foundFavorites))
      .catch(error => this.setState({error: 'Email and password do not match'}))
  }

  clearFields = () => {
    this.setState({ name: '', email: '', password: ''})
  }

  render() {
    const { existingUser, email, name, password, error } = this.state;
    const btnText = existingUser ? 'Sign up' : 'Login'
    return (
      <div className="login-info">
        <button className='login-btn switch-btn' onClick={this.toggleExisting}>{btnText}</button>
        {existingUser &&
      <form className="login-container">
        <p>Login to Favorite Books</p>
        <input
        className='login-input'
        type="text"
        placeholder="Email"
        name="email"
        value={email}
        onChange={this.handleChange} />
        <input
        className='login-input'
        type="password" placeholder="Password"
        name="password"
        value={password}
        onChange={this.handleChange}/>
        <button className='login-btn' onClick={this.loginUser}>Login</button>
        {error && <p className="error-msg">{error}</p>}
      </form>}
      {!existingUser &&
      <form className="login-container">
        <p>Signup to Favorite Books</p>
        <input
        className='login-input'
        type="text"
        placeholder="Name"
        name="name"
        value={name}
        onChange={this.handleChange} />
        <input
        className='login-input'
        type="text"
        placeholder="Email"
        name="email"
        value={email}
        onChange={this.handleChange} />
        <input
        className='login-input'
        type="password"
        placeholder="Create a Password"
        name="password"
        value={password}
        onChange={this.handleChange} />
        <button className='login-btn' onClick={this.submitUser}>Submit</button>
        {error && <p className="error-msg">{error}</p>}
      </form>}
      </div>
    )
  }
};


export const mapDispatchToProps = dispatch => (
  bindActionCreators({ setUser, setFavorites, setGenres }, dispatch)
);

export default connect(null, mapDispatchToProps)(LoginForm);

LoginForm.propTypes = {
  setUser: PropTypes.func,
  setFavorites: PropTypes.func,
  setGenres: PropTypes.func
}
