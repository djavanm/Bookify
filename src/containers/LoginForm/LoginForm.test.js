import React from 'react';
import { shallow } from 'enzyme';
import { LoginForm, mapDispatchToProps } from './LoginForm';
import { setUser, setFavorites, setGenres } from '../../actions';
import { createUser } from '../../util/apiCalls';

jest.mock('../../util/apiCalls');

createUser.mockImplementation(() => {
  return Promise.resolve({
    name: 'Sam',
    email: 'blah@gmail.com',
    password: 'password',
    id: 1
  })
});

describe('LoginForm', () => {
  let wrapper;
  const setUserMock = jest.fn();
  const setFavoritesMock = jest.fn();
  const setGenresMock = jest.fn();
  const mockUser = {
    name: 'Sam',
    email: 'blah@gmail.com',
    password: 'password',
    id: 1
  }

  beforeEach(() => {
    wrapper = shallow(<LoginForm
      setUser={setUserMock}
      setFavorites={setFavoritesMock}
      setGenres={setGenresMock}
      />)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update the state when handle change is called', () => {
    const mockEventName = {
      target: {
        name: 'name',
        value: 'sam'
      }
    };
    const mockEventEmail = {
      target: {
        name: 'email',
        value: 'sam@aol.com'
      }
    };
    const mockEventPassword = {
      target: {
        name: 'password',
        value: 12345
      }
    };
    const expectedName = 'sam';
    const expectedEmail = 'sam@aol.com';
    const expectedPassword = 12345;
     wrapper.instance().handleChange(mockEventName);
     wrapper.instance().handleChange(mockEventEmail);
     wrapper.instance().handleChange(mockEventPassword);
     expect(wrapper.state('name')).toEqual(expectedName);
     expect(wrapper.state('email')).toEqual(expectedEmail);
     expect(wrapper.state('password')).toEqual(expectedPassword);
  });

  it('should toggle existing state boolean after clicking sign up or login button', () => {
    expect(wrapper.state('existingUser')).toEqual(true);
    wrapper.find('.login-btn').at(0).simulate('click');
    expect(wrapper.state('existingUser')).toEqual(false);
  });

  it('should call the createUser fetch when the login or submit user form is submitted', () => {
    const mockEvent = {
      preventDefault: jest.fn()
    }
    wrapper.instance().loginUser(mockEvent);
    expect(createUser).toHaveBeenCalled();
  });

  it('it calls dispatch with the setUser action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = setUser(mockUser);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setUser(mockUser);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
  
});
