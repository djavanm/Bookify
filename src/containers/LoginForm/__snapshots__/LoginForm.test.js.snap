// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match the snapshot 1`] = `
<div>
  <button
    className="login-btn"
    onClick={[Function]}
  >
    Sign up
  </button>
  <form>
    <input
      className="login-input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      className="login-input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <button
      className="login-btn"
      onClick={[Function]}
    >
      Login
    </button>
  </form>
</div>
`;
